Web Inspector: Adding a new style property to a rule after a commented out property results in incorrect style text/bad frontend state

patrickangle opened the pull request:
#### 0d990da3c01519907affff83cfdee71283f2ef4f
<pre>
Web Inspector: Adding a new style property to a rule after a commented out property results in incorrect style text/bad frontend state
<a href="https://bugs.webkit.org/show_bug.cgi?id=247139">https://bugs.webkit.org/show_bug.cgi?id=247139</a>
rdar://88824802

Reviewed by Devin Rousso.

Two problems existed that together caused us to incorrectly associate incoming edits to properties following a
commented-out property with that commented out property, or to create a new WI.CSSProperty instead of using the existing
one that is being edited, resulting in various garbage text being written to the style sheet.

The first issue is in how we handle `pendingProperties` in WI.CSSStyleDeclaration. We currently consider a pending
property to be one that is not enabled (e.g. commented out), which does not match the intent of the comments in the
code and nor does it match its usage. The intended use for `pendingProperties` is to house properties that are no
longer part of the CSSStyleDeclaration after an update, presumedly to allow us to reuse `WI.CSSProperty`s during
undo/redo operations. It is also imaginable that this is where properties being edited should briefly exist if after
their creation, we receive an update from the backend before we have received the update that commits the property under
edit.

The misclassification of what belongs in `pendingProperties` is was not enough by itself to cause the issue however. The
issues arose when DOMNodeStyles._parseStylePropertyPayload attempted to associate incoming property payloads with existing
WI.CSSProperty objects. For pending properties, we only have the property name as a point of verification, which is ripe
for collision when looking at the array of &quot;Pending Properties&quot; that includes commented-out properties.

In addition to properly classifying a &quot;pending&quot; property we need to consider all existing properties for an earlier
comparison between payload properties and WI.CSSProperty objects in WI.DOMNodeStyles.prototype._parseStylePropertyPayload.
We currently are only considering enabled properties for direct matches, even though a commented-out property could
still be a direct match by name and index to an existing property. Failing to make this association results in us
unnecessarily creating a new WI.CSSProperty for the commented out rule, which is turn grows the number of &quot;Pending
Properties&quot; as we toss objects out of the main set of properties for a CSSStyleDeclaration and into the &quot;Pending
Properties&quot;, where they then can end up colliding by name inside WI.DOMNodeStyles.prototype._parseStylePropertyPayload.

We instead need to use all of the existing properties to check for a direct name+index match to catch all possible
properties for reuse.

It seems likely that this regressed sometime during the bringup of the Spreadsheet* editor.
<a href="https://commits.webkit.org/208214@main">https://commits.webkit.org/208214@main</a>, while likely not the cause, renamed the trinity of property arrays as follows:
`properties` -&gt; `enabledProperties`
`allProperties` -&gt; `properties`
`allVisibleProperties` -&gt; `visibleProperties`

It seems ripe for misunderstanding that previously we had both an `allProperties` and a `properties` and I suspect that
the root misunderstanding of this issue stems from that. A quick audit of other uses of `enabledProperties` in our
current code shows generally reasonable usage as far as I can tell.

* LayoutTests/inspector/css/add-css-property-expected.txt:
* LayoutTests/inspector/css/add-css-property.html:

* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype._parseStylePropertyPayload):

Canonical link: <a href="https://commits.webkit.org/256410@main">https://commits.webkit.org/256410@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e49265940d6f04b5c4b22a2403431fca8c61c747

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95486 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4755 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105068 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165336 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4782 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33489 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87856 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100931 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101150 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3495 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82098 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30568 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85387 "Found 1 new API test failure: TestWebKitAPI.FullscreenVideoTextRecognition.TogglePlaybackInElementFullscreen (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87301 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73406 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39254 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36955 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20155 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4433 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40949 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/42807 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42938 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39402 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -217,7 +217,7 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
         }
 
```
This still needs tests, so `merging-blocked` for now.

dcrousso reviewed (approved):
NIT: should "Pending Properties" be capitalized?

> is not enabled (e.g. not commented out)

did you mean

> is not enabled (e.g. commented out)

s/could till/could still

patrickangle committed:
Web Inspector: Adding a new style property to a rule after a commented out property results in incorrect style text/bad frontend state
https://bugs.webkit.org/show_bug.cgi?id=247139
rdar://88824802

Reviewed by Devin Rousso.

Two problems existed that together caused us to incorrectly associate incoming edits to properties following a
commented-out property with that commented out property, or to create a new WI.CSSProperty instead of using the existing
one that is being edited, resulting in various garbage text being written to the style sheet.

The first issue is in how we handle `pendingProperties` in WI.CSSStyleDeclaration. We currently consider a pending
property to be one that is not enabled (e.g. commented out), which does not match the intent of the comments in the
code and nor does it match its usage. The intended use for `pendingProperties` is to house properties that are no
longer part of the CSSStyleDeclaration after an update, presumedly to allow us to reuse `WI.CSSProperty`s during
undo/redo operations. It is also imaginable that this is where properties being edited should briefly exist if after
their creation, we receive an update from the backend before we have received the update that commits the property under
edit.

The misclassification of what belongs in `pendingProperties` is was not enough by itself to cause the issue however. The
issues arose when DOMNodeStyles._parseStylePropertyPayload attempted to associate incoming property payloads with existing
WI.CSSProperty objects. For pending properties, we only have the property name as a point of verification, which is ripe
for collision when looking at the array of "Pending Properties" that includes commented-out properties.

In addition to properly classifying a "pending" property we need to consider all existing properties for an earlier
comparison between payload properties and WI.CSSProperty objects in WI.DOMNodeStyles.prototype._parseStylePropertyPayload.
We currently are only considering enabled properties for direct matches, even though a commented-out property could
still be a direct match by name and index to an existing property. Failing to make this association results in us
unnecessarily creating a new WI.CSSProperty for the commented out rule, which is turn grows the number of "Pending
Properties" as we toss objects out of the main set of properties for a CSSStyleDeclaration and into the "Pending
Properties", where they then can end up colliding by name inside WI.DOMNodeStyles.prototype._parseStylePropertyPayload.

We instead need to use all of the existing properties to check for a direct name+index match to catch all possible
properties for reuse.

It seems likely that this regressed sometime during the bringup of the Spreadsheet* editor.
https://commits.webkit.org/208214@main, while likely not the cause, renamed the trinity of property arrays as follows:
`properties` -> `enabledProperties`
`allProperties` -> `properties`
`allVisibleProperties` -> `visibleProperties`

It seems ripe for misunderstanding that previously we had both an `allProperties` and a `properties` and I suspect that
the root misunderstanding of this issue stems from that. A quick audit of other uses of `enabledProperties` in our
current code shows generally reasonable usage as far as I can tell.

* LayoutTests/inspector/css/add-css-property-expected.txt:
* LayoutTests/inspector/css/add-css-property.html:

* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype._parseStylePropertyPayload):

Canonical link: https://commits.webkit.org/256410@main
